src/hotspot/share/memory/metaspace/metaspaceArena.cpp

```diff
@@ -0,0 +1,493 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "memory/metaspace/allocationGuard.hpp"
+#include "memory/metaspace/chunkManager.hpp"
+#include "memory/metaspace/freeBlocks.hpp"
+#include "memory/metaspace/internalStats.hpp"
+#include "memory/metaspace/metachunk.hpp"
+#include "memory/metaspace/metaspaceArena.hpp"
+#include "memory/metaspace/metaspaceArenaGrowthPolicy.hpp"
+#include "memory/metaspace/metaspaceCommon.hpp"
+#include "memory/metaspace/metaspaceSettings.hpp"
+#include "memory/metaspace/metaspaceStatistics.hpp"
+#include "memory/metaspace/virtualSpaceList.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/init.hpp"
+#include "services/memoryService.hpp"
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+namespace metaspace {
+
+#define LOGFMT         "Arena @" PTR_FORMAT " (%s)"
+#define LOGFMT_ARGS    p2i(this), this->_name
+
+// Returns the level of the next chunk to be added, acc to growth policy.
+chunklevel_t MetaspaceArena::next_chunk_level() const {
+  const int growth_step = _chunks.count();
+  return _growth_policy->get_level_at_step(growth_step);
+}
+
+// Given a chunk, add its remaining free committed space to the free block list.
+void MetaspaceArena::salvage_chunk(Metachunk* c) {
+  if (Settings::handle_deallocations() == false) {
+    return;
+  }
+
+  assert_lock_strong(lock());
+  size_t remaining_words = c->free_below_committed_words();
+  if (remaining_words > FreeBlocks::MinWordSize) {
+
+    UL2(trace, "salvaging chunk " METACHUNK_FULL_FORMAT ".", METACHUNK_FULL_FORMAT_ARGS(c));
+
+    MetaWord* ptr = c->allocate(remaining_words);
+    assert(ptr != NULL, "Should have worked");
+    _total_used_words_counter->increment_by(remaining_words);
+
+    add_allocation_to_fbl(ptr, remaining_words);
+
+    // After this operation: the chunk should have no free committed space left.
+    assert(c->free_below_committed_words() == 0,
+           "Salvaging chunk failed (chunk " METACHUNK_FULL_FORMAT ").",
+           METACHUNK_FULL_FORMAT_ARGS(c));
+  }
+}
```

Comment:
Yes, but in practice, the benefit would be slim:
- it is very rare that you retire a chunk which is not even halfway used. For that to happen a metaspace allocation must be requested larger than half of the current chunk. Does not happen that often.
- This only affects the committed portion of the chunk. So, for larger chunks - where this could hurt - the waste (waste in the sense as "unnecessarily earmarked for one loader only") mostly consists of address space only, not committed space.

Reply:
Okay, I did some statistics and salvaging chunks which are not at least half-way used is super rare. I opt in this case for just leaving the code as it is since splitting the salvaged chunk would mean complicating the code, apart from drawing the central metaspace context lock.

Reply:
Ok. Thanks for doing it.